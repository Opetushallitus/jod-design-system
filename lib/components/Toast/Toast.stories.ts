import type { Meta, StoryObj } from '@storybook/react';

import { Toast } from './Toast';

const meta = {
  title: 'Snackbars/Toast',
  component: Toast,
  tags: ['autodocs'],
} satisfies Meta<typeof Toast>;

export default meta;

type Story = StoryObj<typeof meta>;

const url = 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-8824';
const text = 'Lorem ipsum dolor';

export const Success: Story = {
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is a success toast component for displaying a text.',
      },
    },
  },
  args: {
    text,
    icon: 'check_circle',
    iconAriaLabel: 'Check Circle',
  },
};

export const Warning: Story = {
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is a warning toast component for displaying a text.',
      },
    },
  },
  args: {
    text,
    icon: 'warning',
    iconAriaLabel: 'Warning',
    variant: 'warning',
  },
};

export const ErrorStory: Story = {
  name: 'Error',
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is an error toast component for displaying a text.',
      },
    },
  },
  args: {
    text,
    icon: 'error',
    iconAriaLabel: 'Error',
    variant: 'error',
  },
};

export const Iconless: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542%3A8376',
    },
    docs: {
      description: {
        story: 'This is a toast component for displaying a text without an icon.',
      },
    },
  },
  args: {
    text: 'Lorem ipsum dolor',
  },
};
