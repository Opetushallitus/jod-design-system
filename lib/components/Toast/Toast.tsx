export interface ToastProps {
  /** Text shown on the toast */
  text: string;
  /** Icon shown on the toast */
  icon?: React.ReactNode;
  /** Variant of the toast */
  variant?: 'success' | 'warning' | 'error';
}

/**
 * Toasts display brief, temporary notifications. They are noticeable but do not disrupt the user experience and do not require an action to be taken.
 */
export const Toast = ({ text, icon, variant = 'success' }: ToastProps) => (
  <div
    role="alert"
    aria-live="assertive"
    aria-atomic="true"
    className={`
        ${variant === 'success' ? 'ds-bg-success' : ''}
        ${variant === 'warning' ? 'ds-bg-warning' : ''}
        ${variant === 'error' ? 'ds-bg-alert' : ''}
        ds-inline-flex ds-h-8 ds-items-center ds-gap-3 ds-rounded-sm ds-pl-5 ds-text-black
        ds-text-body-sm ds-font-arial ${icon ? 'ds-pr-7' : 'ds-pr-5'}`
      .replace(/\s+/g, ' ')
      .trim()}
  >
    {icon ?? null}
    {text}
  </div>
);
