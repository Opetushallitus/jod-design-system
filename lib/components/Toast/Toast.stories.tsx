import type { Meta, StoryObj } from '@storybook/react';

import { MdCheck, MdDangerous, MdWarning } from 'react-icons/md';
import { Toast } from './Toast';

const meta = {
  title: 'Snackbar/Toast',
  component: Toast,
  tags: ['autodocs'],
} satisfies Meta<typeof Toast>;

export default meta;

type Story = StoryObj<typeof meta>;

const url = 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-8824';
const text = 'Lorem ipsum dolor';

export const Success: Story = {
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is a success toast component for displaying a text.',
      },
    },
  },
  args: {
    text,
    icon: <MdCheck size={24} />,
  },
};

export const Warning: Story = {
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is a warning toast component for displaying a text.',
      },
    },
  },
  args: {
    text,
    icon: <MdWarning size={24} />,
    variant: 'warning',
  },
};

export const ErrorStory: Story = {
  name: 'Error',
  parameters: {
    design: {
      type: 'figma',
      url,
    },
    docs: {
      description: {
        story: 'This is an error toast component for displaying a text.',
      },
    },
  },
  args: {
    text,
    icon: <MdDangerous size={24} />,
    variant: 'error',
  },
};

export const Iconless: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542%3A8376',
    },
    docs: {
      description: {
        story: 'This is a toast component for displaying a text without an icon.',
      },
    },
  },
  args: {
    text: 'Lorem ipsum dolor',
  },
};
