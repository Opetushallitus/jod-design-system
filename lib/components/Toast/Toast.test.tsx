import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

import { Toast } from './Toast';

describe('Toast', () => {
  it('renders the toast component with text', () => {
    const { container } = render(<Toast text="This is a toast message" />);
    const toastElement = screen.getByRole('alert');
    expect(toastElement).toBeInTheDocument();
    expect(toastElement).toHaveTextContent('This is a toast message');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the toast component with icon', () => {
    const { container } = render(<Toast text="This is a toast message" icon="mood" />);
    const toastElement = screen.getByRole('alert');
    expect(toastElement).toBeInTheDocument();
    expect(toastElement).toHaveTextContent('This is a toast message');
    expect(container.firstChild).toMatchSnapshot();
  });
});
