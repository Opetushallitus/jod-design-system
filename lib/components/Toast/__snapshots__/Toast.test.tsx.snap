// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toast > renders the toast component with icon 1`] = `
<div
  aria-atomic="true"
  aria-live="assertive"
  class="bg-success text-white inline-flex h-8 items-center gap-3 rounded-[4px] pl-5 text-body-sm pr-7"
  role="alert"
>
  <span
    aria-label="Mood"
    class="material-symbols-outlined size-24 select-none"
    role="img"
  >
    mood
  </span>
  This is a toast message
</div>
`;

exports[`Toast > renders the toast component with text 1`] = `
<div
  aria-atomic="true"
  aria-live="assertive"
  class="bg-success text-white inline-flex h-8 items-center gap-3 rounded-[4px] pl-5 text-body-sm pr-5"
  role="alert"
>
  This is a toast message
</div>
`;
